File: autovideo/api/routes/bgm.py
from __future__ import annotations

from typing import Any, NoReturn

from fastapi import APIRouter, Depends, File, Form, Request, Response, UploadFile, status
from fastapi.responses import FileResponse
from pydantic import BaseModel, Field

from autovideo.api.dependencies import get_settings
from autovideo.api.errors import structured_error
from autovideo.core.settings import Settings
from autovideo.services.bgm import (
    BgmCategoryDuplicateError,
    BgmCategoryEmptyError,
    BgmCategoryNameRequiredError,
    BgmCategoryNotFoundError,
    BgmFileEmptyError,
    BgmFileTooLargeError,
    BgmFileUnsupportedError,
    BgmLibraryCorruptError,
    BgmLibraryService,
    BgmTrackFileDeleteError,
    BgmTrackNameRequiredError,
    BgmTrackNotFoundError,
)
from autovideo.services.bgm.models import SUPPORTED_BGM_EXTENSIONS

router = APIRouter(prefix="/api/bgm", tags=["bgm"])


class BgmCategoryRequest(BaseModel):
    name: str = Field(min_length=1, max_length=80)


class BgmTrackUpdateRequest(BaseModel):
    display_name: str = Field(min_length=1, max_length=120)
    category_id: str | None = None


def _service(request: Request, settings: Settings) -> BgmLibraryService:
    return BgmLibraryService(
        settings,
        audio_probe=getattr(request.app.state, "bgm_audio_probe", None),
    )


def _public_library(library: dict[str, Any]) -> dict[str, Any]:
    items = library.get("items")
    categories = library.get("categories")
    public_items = items if isinstance(items, list) else []
    public_categories = categories if isinstance(categories, list) else []
    return {
        "items": public_items,
        "categories": public_categories,
        "total_tracks": int(library.get("total_tracks") or len(public_items)),
        "storage_status": "ready",
        "supported_extensions": sorted(SUPPORTED_BGM_EXTENSIONS),
    }


def _raise_bgm_error(exc: Exception) -> NoReturn:
    if isinstance(exc, BgmFileUnsupportedError):
        raise structured_error(
            status.HTTP_400_BAD_REQUEST,
            "BGM_FILE_UNSUPPORTED",
            allowed=sorted(SUPPORTED_BGM_EXTENSIONS),
        ) from exc
    if isinstance(exc, BgmFileEmptyError):
        raise structured_error(status.HTTP_400_BAD_REQUEST, "BGM_FILE_EMPTY") from exc
    if isinstance(exc, BgmFileTooLargeError):
        raise structured_error(
            status.HTTP_413_CONTENT_TOO_LARGE,
            "BGM_FILE_TOO_LARGE",
            max_upload_bytes=exc.max_upload_bytes,
        ) from exc
    if isinstance(exc, BgmTrackNotFoundError):
        raise structured_error(status.HTTP_404_NOT_FOUND, "BGM_TRACK_NOT_FOUND") from exc
    if isinstance(exc, BgmCategoryNotFoundError):
        raise structured_error(status.HTTP_404_NOT_FOUND, "BGM_CATEGORY_NOT_FOUND") from exc
    if isinstance(exc, BgmCategoryDuplicateError):
        raise structured_error(status.HTTP_400_BAD_REQUEST, "BGM_CATEGORY_DUPLICATE") from exc
    if isinstance(exc, BgmCategoryNameRequiredError):
        raise structured_error(status.HTTP_400_BAD_REQUEST, "BGM_CATEGORY_NAME_REQUIRED") from exc
    if isinstance(exc, BgmTrackNameRequiredError):
        raise structured_error(status.HTTP_400_BAD_REQUEST, "BGM_TRACK_NAME_REQUIRED") from exc
    if isinstance(exc, BgmCategoryEmptyError):
        raise structured_error(status.HTTP_400_BAD_REQUEST, "BGM_CATEGORY_EMPTY") from exc
    if isinstance(exc, BgmLibraryCorruptError):
        raise structured_error(status.HTTP_500_INTERNAL_SERVER_ERROR, "BGM_LIBRARY_CORRUPT") from exc
    if isinstance(exc, BgmTrackFileDeleteError):
        raise structured_error(
            status.HTTP_500_INTERNAL_SERVER_ERROR,
            "BGM_TRACK_FILE_DELETE_FAILED",
        ) from exc
    raise exc


@router.get("")
def list_bgm(request: Request, settings: Settings = Depends(get_settings)) -> dict[str, Any]:
    try:
        return _public_library(_service(request, settings).library())
    except Exception as exc:
        _raise_bgm_error(exc)


@router.post("/tracks", status_code=status.HTTP_201_CREATED)
async def upload_bgm_track(
    request: Request,
    file: UploadFile = File(...),
    category_id: str | None = Form(default=None),
    settings: Settings = Depends(get_settings),
) -> dict[str, Any]:
    try:
        return _service(request, settings).store_track(
            content=await file.read(),
            original_filename=file.filename or "bgm",
            category_id=category_id or None,
        )
    except Exception as exc:
        _raise_bgm_error(exc)


@router.put("/tracks/{track_id}")
def update_bgm_track(
    track_id: str,
    body: BgmTrackUpdateRequest,
    request: Request,
    settings: Settings = Depends(get_settings),
) -> dict[str, Any]:
    try:
        return _service(request, settings).update_track(
            track_id,
            display_name=body.display_name,
            category_id=body.category_id,
        )
    except Exception as exc:
        _raise_bgm_error(exc)


@router.delete("/tracks/{track_id}")
def delete_bgm_track(
    track_id: str,
    request: Request,
    settings: Settings = Depends(get_settings),
) -> dict[str, Any]:
    try:
        return _service(request, settings).delete_track(track_id)
    except Exception as exc:
        _raise_bgm_error(exc)


@router.get("/tracks/{track_id}/file")
def download_bgm_track(
    track_id: str,
    request: Request,
    settings: Settings = Depends(get_settings),
) -> Response:
    try:
        path, media_type, filename = _service(request, settings).track_file(track_id)
    except Exception as exc:
        _raise_bgm_error(exc)
    response = FileResponse(path, media_type=media_type, filename=filename)
    response.headers["X-Content-Type-Options"] = "nosniff"
    return response


@router.post("/categories", status_code=status.HTTP_201_CREATED)
def create_bgm_category(
    body: BgmCategoryRequest,
    request: Request,
    settings: Settings = Depends(get_settings),
) -> dict[str, Any]:
    try:
        return _service(request, settings).create_category(body.name)
    except Exception as exc:
        _raise_bgm_error(exc)


@router.put("/categories/{category_id}")
def update_bgm_category(
    category_id: str,
    body: BgmCategoryRequest,
    request: Request,
    settings: Settings = Depends(get_settings),
) -> dict[str, Any]:
    try:
        return _service(request, settings).update_category(category_id, body.name)
    except Exception as exc:
        _raise_bgm_error(exc)


@router.delete("/categories/{category_id}")
def delete_bgm_category(
    category_id: str,
    request: Request,
    settings: Settings = Depends(get_settings),
) -> dict[str, Any]:
    try:
        return _service(request, settings).delete_category(category_id)
    except Exception as exc:
        _raise_bgm_error(exc)

File: frontend/src/api/bgm.ts
export interface BgmCategory {
  id: string;
  name: string;
  sort_order: number;
  track_count: number;
  created_at: string;
  updated_at: string;
}

export interface BgmTrack {
  id: string;
  filename: string;
  original_filename: string;
  display_name: string;
  category_id: string | null;
  category_name: string;
  media_type: string;
  extension: string;
  size_bytes: number;
  duration_seconds: number;
  audio_url: string;
  created_at: string;
  updated_at: string;
}

export interface BgmLibrary {
  items: BgmTrack[];
  categories: BgmCategory[];
  storage_status: "ready";
  total_tracks: number;
  supported_extensions: string[];
}

export class BgmApiError extends Error {
  readonly code: string;
  readonly status: number;
  readonly detail: Record<string, unknown>;

  constructor(code: string, status: number, detail: Record<string, unknown> = {}) {
    super(code);
    this.name = "BgmApiError";
    this.code = code;
    this.status = status;
    this.detail = detail;
  }
}

function responseErrorCode(payload: unknown, status: number): string {
  if (
    typeof payload === "object" &&
    payload !== null &&
    "detail" in payload &&
    typeof payload.detail === "object" &&
    payload.detail !== null &&
    "code" in payload.detail &&
    typeof payload.detail.code === "string"
  ) {
    return payload.detail.code;
  }
  return `HTTP_${status}`;
}

async function readJson<T>(response: Response): Promise<T> {
  if (!response.ok) {
    const payload = await response.json().catch(() => ({}));
    const detail =
      typeof payload === "object" &&
      payload !== null &&
      "detail" in payload &&
      typeof payload.detail === "object" &&
      payload.detail !== null
        ? (payload.detail as Record<string, unknown>)
        : {};
    throw new BgmApiError(responseErrorCode(payload, response.status), response.status, detail);
  }
  return response.json() as Promise<T>;
}

function detailNumber(error: BgmApiError, key: string): number | null {
  const value = error.detail[key];
  return typeof value === "number" ? value : null;
}

export function readableBgmError(error: unknown): string {
  if (error instanceof BgmApiError) {
    if (error.code === "BGM_FILE_UNSUPPORTED") {
      return "不支持的音频格式或文件中没有音频流";
    }
    if (error.code === "BGM_FILE_EMPTY") {
      return "上传的 BGM 文件为空";
    }
    if (error.code === "BGM_FILE_TOO_LARGE") {
      const limit = detailNumber(error, "max_upload_bytes");
      return limit ? `BGM 文件超过上传大小限制（${limit} 字节）` : "BGM 文件超过上传大小限制";
    }
    if (error.code === "REQUEST_TOO_LARGE") {
      const limit = detailNumber(error, "max_request_bytes");
      return limit ? `BGM 文件超过请求大小限制（${limit} 字节）` : "BGM 文件超过上传大小限制";
    }
    if (error.code === "BGM_TRACK_NOT_FOUND") {
      return "当前 BGM 不存在或已删除";
    }
    if (error.code === "BGM_CATEGORY_NOT_FOUND") {
      return "当前 BGM 分类不存在或已删除";
    }
    if (error.code === "BGM_CATEGORY_DUPLICATE") {
      return "分类名已存在";
    }
    if (error.code === "BGM_CATEGORY_NAME_REQUIRED") {
      return "请输入分类名";
    }
    if (error.code === "BGM_TRACK_NAME_REQUIRED") {
      return "请输入 BGM 名称";
    }
    if (error.code === "BGM_CATEGORY_EMPTY") {
      return "这个分类下没有可用 BGM";
    }
    if (error.code === "BGM_LIBRARY_CORRUPT") {
      return "BGM 库元数据异常，请检查数据文件";
    }
    if (error.code === "BGM_TRACK_FILE_DELETE_FAILED") {
      return "BGM 文件删除失败，请检查文件权限后重试";
    }
  }
  return error instanceof Error ? error.message : "BGM 操作失败";
}

export async function fetchBgmLibrary(): Promise<BgmLibrary> {
  return readJson(await fetch("/api/bgm"));
}

export async function uploadBgmTrack(input: {
  file: File;
  category_id?: string | null;
}): Promise<BgmTrack> {
  const formData = new FormData();
  formData.append("file", input.file);
  if (input.category_id) {
    formData.append("category_id", input.category_id);
  }
  return readJson(await fetch("/api/bgm/tracks", { method: "POST", body: formData }));
}

export async function updateBgmTrack(input: {
  id: string;
  display_name: string;
  category_id?: string | null;
}): Promise<BgmTrack> {
  const body: { display_name: string; category_id?: string | null } = {
    display_name: input.display_name,
  };
  if (input.category_id !== undefined) {
    body.category_id = input.category_id;
  }

  return readJson(
    await fetch(`/api/bgm/tracks/${encodeURIComponent(input.id)}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }),
  );
}

export async function deleteBgmTrack(trackId: string): Promise<{ id: string; deleted: boolean }> {
  return readJson(
    await fetch(`/api/bgm/tracks/${encodeURIComponent(trackId)}`, { method: "DELETE" }),
  );
}

export async function createBgmCategory(input: { name: string }): Promise<BgmCategory> {
  return readJson(
    await fetch("/api/bgm/categories", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(input),
    }),
  );
}

export async function updateBgmCategory(input: {
  id: string;
  name: string;
}): Promise<BgmCategory> {
  return readJson(
    await fetch(`/api/bgm/categories/${encodeURIComponent(input.id)}`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ name: input.name }),
    }),
  );
}

export async function deleteBgmCategory(
  categoryId: string,
): Promise<{ id: string; deleted: boolean }> {
  return readJson(
    await fetch(`/api/bgm/categories/${encodeURIComponent(categoryId)}`, { method: "DELETE" }),
  );
}
